rust/ql/test/query-tests/security/CWE-770/main.rs

```diff
@@ -210,7 +210,7 @@ unsafe fn test_system_alloc(v: usize) {
             let _ = std::alloc::System.grow_zeroed(m4, l4, l2).unwrap(); // $ Alert[rust/uncontrolled-allocation-size]=arg1
         }
     } else {
-        let _ = std::alloc::System.shrink(m4, l4, l2).unwrap();
+        let _ = std::alloc::System.shrink(m4, l4, l2).unwrap(); // $ SPURIOUS: Alert[rust/uncontrolled-allocation-size]=arg1 - FP
```

Comment:
I think the model generator is right to generate flow for this, `shrink` is causing an allocation to happen and the new size `l2` is user controlled.  The reason this is a FP is because we know a `shrink` can only reduce the size of the allocation so this can be the source of a problem.

I think the right answer is to introduce a barrier.  I can do this as follow-up to your PR.

Reply:
well, reading https://doc.rust-lang.org/beta/std/alloc/trait.Allocator.html#safety-4 I can't decide whether we can count on shrink actually shrinking, considering that the constraint of `l2`'s size being smaller that `l4`'s one is actually a safety contract to be guaranteed by the user.

However apart from that, we should know that `l2` has size `v`, and that `v < 10` here, which means we aren't really unconstrained in any case here. Or is this too much for the current implementation of the analysis? Are we doing some kind of range analysis?

Reply:
We don't have range analysis, we do have barrier guards and in fact the query is using them.  However in this case the check on `v` is _after_ the `std::alloc::Layout` has been created from it, so we don't notice the guard applies to the layout.  I'll see if I can easily address this...